web/src/app/services/data-store/data-store.service.ts

```
@@ -249,39 +254,71 @@ export class DataStoreService {
+      .where(l.jobId, '==', jobId);
 
-    return await this.db
-      .collection(SURVEYS_COLLECTION_NAME)
-      .doc(surveyId)
-      .delete();
-  }
+    const loisSnapshot = await loisQuery.get();
+    loisSnapshot.forEach(doc => {
+      transaction.delete(doc.ref);
+    });
 
-  async deleteJob(surveyId: string, jobId: string) {
-    await this.deleteAllLocationsOfInterestInJob(surveyId, jobId);
-    await this.deleteAllSubmissionsInJob(surveyId, jobId);
-    return await this.db
+    const submissionsQuery = firestore
+      .collection(`${SURVEYS_COLLECTION_NAME}/${surveyId}/submissions`)
+      .where(sb.jobId, '==', jobId);
+
+    const submissionsSnapshot = await submissionsQuery.get();
+    submissionsSnapshot.forEach(doc => {
+      transaction.delete(doc.ref);
+    });
+
+    const jobRef = firestore
       .collection(`${SURVEYS_COLLECTION_NAME}/${surveyId}/jobs`)
-      .doc(jobId)
-      .delete();
+      .doc(jobId);
+    transaction.delete(jobRef);
   }
 
-  private async deleteAllSubmissionsInJob(surveyId: string, jobId: string) {
-    const submissions = this.db.collection(
-      `${SURVEYS_COLLECTION_NAME}/${surveyId}/submissions`,
-      ref => ref.where(sb.jobId, '==', jobId)
-    );
-    const querySnapshot = await firstValueFrom(submissions.get());
-    return await Promise.all(querySnapshot.docs.map(doc => doc.ref.delete()));
+  async deleteSurvey(survey: Survey): Promise<void> {
+    const {id: surveyId, jobs} = survey;
+
+    try {
+      await this.db.firestore.runTransaction(async transaction => {
+        for (const {id: jobId} of jobs.values()) {
+          await this._deleteJobAndRelatedData(transaction, surveyId, jobId);
+        }
+
+        const surveyRef = this.db.firestore
+          .collection(SURVEYS_COLLECTION_NAME)
+          .doc(surveyId);
+
+        transaction.delete(surveyRef);
+      });
+    } catch (error) {
+      console.error(`Error deleting survey: ${surveyId}`, error);
+      throw error;
```

If we log and rethrow we'll end up with duplicate log messages which will add noise and make debugging harder. Why not just let the error throw and let the exception will be logged upstream?